__tests__/frontend-proxy.ts

```diff
@@ -0,0 +1,215 @@
+import {
+  handleRequest,
+  FRONTEND_DOMAIN,
+  createApiQuery,
+} from '../src/frontend-proxy'
+import { createJsonResponse } from '../src/utils'
+import { withMockedFetch, hasOkStatus } from './utils'
+
+const TEST_ALLOWED_TYPES = ['User', 'Article']
+const testUrl = 'https://de.serlo.org/_nextexample-url'
+
+describe('handleRequest()', () => {
+  describe('returns null if language tenant is not "de"', () => {
+    test.each([
+      'https://serlo.org/',
+      'http://ta.serlo.org/',
+      'https://stats.serlo.org/',
+      'http://en.serlo.org/math',
+    ])('URL is %p', async (url) => {
+      expect(await handleUrl(url)).toBeNull()
+    })
+  })
+
+  describe('chooses backend based on probability', () => {
+    test.each([
+      [1, TEST_ALLOWED_TYPES[0], 'frontend', true],
+      [1, TEST_ALLOWED_TYPES[1], 'frontend', true],
+      [0, TEST_ALLOWED_TYPES[0], 'no-frontend', false],
+      [0, TEST_ALLOWED_TYPES[1], 'no-frontend', false],
+    ])(
+      'probability=%p',
+      async (probability, typename, responseText, useFrontendCookie) => {
+        await withMockedFetch(
+          [createApiResponse(typename), checkFrontendRequest],
+          async () => {
+            const res = await handleUrl(testUrl, probability)
+
+            expect(await res.text()).toBe(responseText)
+            expect(res.headers.get('Set-Cookie')).toBe(
+              `useFrontend${probability * 100}=${useFrontendCookie}; path=/`
+            )
+          }
+        )
+      }
+    )
```

Comment:
I don't really like this test:
- The `test.each` makes it hard to read. I have to jump between the values, their names and their usage.
- I also have to jump to the top if I want to know what `testUrl` or `TEST_ALLOWED_TYPES` stand for.
- It doesn't feel like checking different types improves the test since the tested assertion here should completely independent of the type, right? If so, I'd argue that you can just choose one type and be done with it (making the test more readable and emphasizing what the actual difference between the test cases is).
- `createApiResponse` & `checkFrontendRequest` are two additional helpers that I need to check to understand this test. `createApiResponse` is probably okay since the naming makes sense, but I don't know what `checkFrontendRequest` is supposed to do, especially as it depends on request and contains logic?
- `withMockedFetch` is another clever helper that I don't really like in the tests. Again, I have to check its implementation (that might be error-prone btw) to understand what it is doing. What does the array represent? Is the order relevant? Why does it accept both responses and functions that create the response? IMHO: inlining the `withMockedFetch` calls does make this way more readable and I only have to understand `jest` to get the gist of what the test wants to achieve.
- Why do we call `handleUrl` when this is another function in the tests that has some backend logic? Why not make the thing that it does explicit and call the production code directly? I shouldn't need to jump to the implementation to see that `handleUrl` apparently also (sometimes?) creates a Cookie before calling the backend logic. That should be part of the test.

What I'd like to strive for in the tests instead is:
- Make it as easy as possible to understand what the tests achieves to check. I want to understand what the test has as assumptions (given), what the tests executes (when) and what it checks (then). In this specific example, given & when are hidden inside `handleUrl` and `withMockedFetch` with some logic behind it (do you assume that the user already has the cookie or not?), assertions are also kinda over the place (in `withMockedFetch` and its callback)
- I'd like to understand what the edge cases are. Did you check `User` and `Article` because they have a different behavior that we need to watch out for? There is no value to checking everything. A test is not a proof but rather a concrete example usage that represents a set of inputs. So I'd rather have one example for the good case + one example each for relevant edge cases. Having more than two tests in the same set of inputs just makes the tests harder to read and increases the time it takes to run the tests.
- Tests should be self-explanatory as possible. If there is too much duplicate code so that we create helpers, helpers should be as "dumb" as possible (no own logic) and restrict themselves to one part of the test (given / when or then). It might even be possible to move common setup into a `beforeEach` and forgo with the helper.

Reply:
```
const fetch = jest.fn().mockReturnValueOnce(value)
global.fetch = fetch
fetch.mockImplementationOnce(createApiResponse(typename))

const res = await handleUrl(testUrl, probability)
expectFetchToHaveBeenCalledWithRequest({
    url: 'frontend.serlo.org', // Or whatever,
    headers: {
        'Set-Cookie': `useFrontend${probability * 100}=${useFrontendCookie}; path=/`
    }
})
```

Reply:
- [ ] more 2 argumente -> object argument
- [ ] delete TEST_URLS and TEST_ALLOWED_TYPES